lib/datadog/tracing/contrib/rack/middlewares.rb

```diff
@@ -134,6 +134,11 @@ def set_request_tags!(trace, request_span, env, status, headers, response, origi
             # Otherwise, the getter method would delegate to its root span
             trace.resource = request_span.resource unless trace.resource_override?
 
+            # We add this metric when ASM standalone is enabled to make sure we don't bill APM
+            if Datadog.configuration.appsec.standalone.enabled
+              request_span.set_metric(Tracing::Metadata::Ext::TAG_APM_ENABLED, 0)
+            end
```

Comment:
In https://github.com/DataDog/dd-trace-rb/pull/3965 @anmarchenko mentioned tracing gaining a dependency on appsec and that pattern appears to continue here. I my opinion it would be better for tracing to provide hooks that appsec will hook into to avoid tracing referencing appsec in code, but I am not developing either part of the library.

I do also find it hard to ascertain whether the code behaves correctly - this PR sets APM enabled to 0 from tracing, but it is not clear from the diff where/under what conditions APM enabled is set to 1 (or if that happens at all).

Reply:
Can you elaborate the reason about tagging a `rack` span specifically?

Reply:
The env var name `DD_EXPERIMENTAL_APPSEC_STANDALONE_ENABLED` is temporary, and what Standalone Appsec actually does is disabling tracing (or more precisely rate-limiting it at 1 trace per minute), and when 'Appsec' is activated along 'Standalone Appsec' (as Standalone Appsec does not in fact activate Appsec by itself), it force-keep traces that contains Appsec events.

There is a RFC currently being reviewed to find the final name for this env var, which will most likely be `DD_TRACE_APM_ENABLED`. When it will be approved, I will also change the configuration from `Datadog.configuration.appsec.standalone.enabled` to something like `Datadog.configuration.tracing.apm.enabled`, which will make more sense and show that there is no dependency from appsec in tracing (but the other way will be true, which already is anyway)

Reply:
> Can you elaborate the reason about tagging a rack span specifically?

For now, the only product other than tracing that uses traces is Appsec, and in Appsec we only instrument rack-based apps (Rails, Sinatra, Rack itself...)
But maybe we could tag somewhere else, like in http frameworks, now that we have moved this part in tracing.

Reply:
I understand this is experimental feature but similar concern with shotgun surgery and product coupling has been raised in the [previous pull request code review](https://github.com/DataDog/dd-trace-rb/pull/3965#discussion_r1804435163), that signals the severity of design issue. 

Currently, 
1. It is hard to make sense of the convention and logic using `Datadog.configuration.tracing.apm.enabled`, which APM includes multiple products including `profiling`. 
2. I can see similar approach like `rack`, that AppSec contains the 3rd party patches that depends on TraceOperation.
3. Or maybe other kinds of dependency injection mechanism

Overall, the requirements (some kinds of hook systems, perhaps could be encompassed by @lloeki 's Instrumentation API ?) are not clear and I am afraid that if we are indifferent now could cast a long shadow in the future.

Reply:
does it make sense to find the root span here and not the rack span?

Reply:
Yes, this is changed in e0c503a06a8d14b82701e51ca0a0cef8dc2167d2

Note that "non-billing" naming is temporary.